### PR TITLE
Add QUIC support to s_client

### PR DESCRIPTION
**Pull request by mattcaswell**
This PR adds support to s_client for creating QUIC connections through the new flag "-quic". It also adds a new command for signalling FIN on a stream.

It is built on top of #20566 and includes the commits from there. Only the last 2 commits are relevant to this PR.

### CODE REVIEW

**Comment by mattcaswell** [comment]
This is draft because it depends on #20566. It also needs some documentation adding.

**Comment by mattcaswell** [comment]
I've rebased this onto the latest version of #20566

**Comment by mattcaswell** [comment]
Using this PR you can use s_client to connect to the picoquic sample server.

To do that clone the picoquic repo from here:

https://github.com/private-octopus/picoquic

Build picoquic according to the instructions in the README.md file.

Note you need a *very* recent version of picoquic for this to work. Must include commit 6af4c8c6 or above (which fixes a bug in the picoquic sample server).

From the root of the picoquic directory create some sample data, e.g.

```` bash
$ mkdir samplefiles
$ echo "Helloworld" >samplefiles/sample.txt
````

Now start the picoquic sample server:

```` bash
$ ./picoquic_sample server 9999 certs/cert.pem certs/key.pem samplefiles
````

And execute s_client build from this PR like this:

```` bash
$ openssl s_client -quic -alpn picoquic_sample -connect localhost:9999 -adv
````

This should connect successfully to the picoquic_sample server and enter the new s_client advanced mode. You don't strictly need advanced mode to get this far - but basic mode doesn't offer the ability to indicate FIN on a stream which you need for the next step.

From the s_client session enter the following:

````
sample.txt{fin}
````

This will request the contents of the sample.txt file from the server and then immediately signal FIN. You should see the contents of the file echoed into the s_client terminal. To close the connection type this into s_client:

````
{quit}
````

**Comment by mattcaswell** [comment]
Now that #20566 has been merged I now consider this PR ready for review, so I have taken it out of draft.

Please take a look.

File: apps/s_client.c
Comment (t-j-h) [inline apps/s_client.c:3083]: SSL_tick seems poorly named ...

File: apps/s_client.c
Comment (t-j-h) [inline apps/s_client.c:112]: Adding SSL_is_quic() as a separate PR first would make sense to me.
If that was done perhaps an SSL_is_tls() or perhaps a generic function to return the protocol class/type as an integer.

Internally we have SSL_CONNECTION_IS_TLS13(), SSL_CONNECTION_IS_QUIC() and SSL_CONNECTION_IS_DTLS()

SSL_get_version() is "close" to this and what many applications use to report information - but that won't return anything useful in the context of QUIC and perhaps it should.

File: apps/s_client.c
Comment (t-j-h) [inline apps/s_client.c:1379]: A good question. What happens to APIs and values for the next version of QUIC?

File: apps/s_client.c
Comment (mattcaswell) [reply apps/s_client.c:112]: I had already come to the same conclusion and hence #20650.

File: apps/s_client.c
Comment (mattcaswell) [reply apps/s_client.c:2921]: @hlandau - thoughts?

Comment (hlandau) [reply apps/s_client.c:2921]: I designed `SSL_get_tick_timeout` like this because `DTLSv1_get_tick_timeout` doesn't make a distinction between a failed call and an infinity value. In general I'd like to avoid adding new APIs which can't indicate a failure condition.

File: apps/s_client.c
Comment (mattcaswell) [reply apps/s_client.c]: Now that #20650 has been merged I've updated this PR to use it and removed this TODO.

File: apps/s_client.c
Comment (mattcaswell) [reply apps/s_client.c]: Initial updates for QUIC version handling were made in #20650. How that applies to min/max version handling was extensively debated in #20790 and subsequently in this week's OTC. The result is #20830. I've updated this PR to be in line with #20830.

File: apps/s_client.c
Comment (mattcaswell) [reply apps/s_client.c:2914]: @t-j-h what is your view on @hlandau's response?

Either way, if this requires further work, I think it should be handled by a different issue and not hold up this PR.

Comment (t-j-h) [reply apps/s_client.c:2914]: Signalling there is no active timeout isn't an "error". A separate call can be used for that. Using negative value indicators and have variations across protocols is asking for confusion. Be more explicit in the handling. This is what leads to -2 and -3 return values getting added.

File: apps/s_client.c
Comment (mattcaswell) [reply apps/s_client.c:3076]: The SSL_tick() name was first proposed and discussed a long time ago. I'm personally fine with the name. But if we want to change it then we should probably raise a separate issue for it and discuss it in OTC. I don't think it should hold up this PR.

@t-j-h - should I raise an issue for this?

**Comment by mattcaswell** [comment]
I've rebased this and either addressed or otherwise responded to all of the feedback. Please take another look.

File: apps/s_client.c
Comment (t-j-h) [reply apps/s_client.c:3076]: Let's get an OTC decision on the API related items so we have clear direction and can apply whatever the agreed principles end up being. Seeing it in use should help crystalise the views rapidly.

**Comment by mattcaswell** [comment]
** OTC Question: Should SSL_tick() be renamed? **

** OTC Question: Is signalling no active timeout by setting tv->tv_sec to -1 in SSL_get_tick_timeout() the correct approach? **

File: apps/s_client.c
Comment (mattcaswell) [reply apps/s_client.c:3076]: I've put an OTC hold on the PR to discuss it.

**Comment by paulidale** [comment]
OTC: rename `SSL_tick()` to `SSL_process_events()` or `SSL_handle_events()`.  Also rename `SSL_get_tick_timeout()` appropriately.  Finally, fix all the documentation.  Issue: #20867

OTC: add third parameter to `SSL_get_tick_timeout()` indicating lack of timeout.  Issue: #20868

This PR isn't blocked by either of these.

File: apps/s_client.c
Comment (hlandau) [reply apps/s_client.c:2914]: Marking this as resolved as it is sorted in #20867.

File: apps/s_client.c
Comment (hlandau) [reply apps/s_client.c:3076]: Marking this as resolved as it is sorted in https://github.com/openssl/openssl/issues/20867.

**Review by hlandau — APPROVED** [review]

**Comment by hlandau** [comment]
The SSL_tick rename is done in #20867. If this PR merges before that PR, I'll rebase and update it accordingly. So not a blocker for this PR.

**Comment by mattcaswell** [comment]
Ping for second review.

**Review by t-j-h — APPROVED** [review]

**Comment by mattcaswell** [comment]
Pushed. Thanks.

**Pull request closed without merging**
